# Ouvrir les fichiers CSV avec Pandas

In [1]:
import pandas as pd
import numpy as np
#pd.set_option("max_rows", None)  #Préciser le nombre de lignes max


In [2]:
df_population = pd.read_csv("old_data/fr_population.csv")
df_animaux = pd.read_csv("old_data/fr_animaux.csv")
df_vegetaux = pd.read_csv("old_data/fr_vegetaux.csv")

# Réponse 1  : Nombre total d’humains sur la planète

In [3]:
df_population[df_population['Code année']==2013]['Valeur'].sum()

8413993

### ** Le nombre total en 2013 est 8413993 * 1000 personnes **


### Vérifier les valeurs manquantes ou si une année se répète 2 fois

In [4]:
df_population['Code année'].isna().sum(),df_population['Année'].isna().sum()

(0, 0)

In [5]:
df_population[df_population['Code année']==2013]['Zone'].value_counts()  #calculer le nombre de valeur qui se répètent selon le code du pays

Afghanistan    1
Oman           1
Namibie        1
Népal          1
Nicaragua      1
              ..
France         1
Gabon          1
Gambie         1
Géorgie        1
Zimbabwe       1
Name: Zone, Length: 175, dtype: int64

# Réponse 2 : Equation à 3 termes

#### Disponibilité intérieure = Production + Importations + |Variation de stock| - Exportations = 
#### Aliments pour animaux + Semences + Pertes + Traitement + Autres utilisations (non alimentaire) + Nourriture


## Exemple blé en france 2018

In [6]:
# Données récupérées à partir du lien : https://www.fao.org/faostat/fr/#data/FBS
data_ble_france2018= pd.read_csv('old_data/ble_france_2018.csv')

In [7]:
data_ble_france2018.head(3)

Code Domaine                      Domaine  Code zone (FAO)    Zone  \
0          FBS  Bilans Alimentaires (2014-)               68  France   
1          FBS  Bilans Alimentaires (2014-)               68  France   
2          FBS  Bilans Alimentaires (2014-)               68  France   

   Code Élément                  Élément  Code Produit          Produit  \
0          5511               Production          2511  Blé et produits   
1          5611  Importations - Quantité          2511  Blé et produits   
2          5072       Variation de stock          2511  Blé et produits   

   Code année  Année               Unité  Valeur Symbole  \
0        2018   2018  Milliers de tonnes   35798      Im   
1        2018   2018  Milliers de tonnes    2349      Im   
2        2018   2018  Milliers de tonnes   -1819      Im   

                              Description du Symbole  
0  Données de la FAO basées sur une méthodologie ...  
1  Données de la FAO basées sur une méthodologie ...  
2  Données de la FAO basées sur une méthodologie ...

In [8]:
data_ble_france2018 = data_ble_france2018[['Code Élément', 'Élément', 'Unité', 'Valeur']]

In [9]:
data_ble_france2018

Code Élément                                Élément               Unité  \
0           5511                             Production  Milliers de tonnes   
1           5611                Importations - Quantité  Milliers de tonnes   
2           5072                     Variation de stock  Milliers de tonnes   
3           5911                Exportations - Quantité  Milliers de tonnes   
4           5301               Disponibilité intérieure  Milliers de tonnes   
5           5521                  Aliments pour animaux  Milliers de tonnes   
6           5527                               Semences  Milliers de tonnes   
7           5123                                 Pertes  Milliers de tonnes   
8           5131                             Traitement  Milliers de tonnes   
9           5154  Autres utilisations (non alimentaire)  Milliers de tonnes   
10          5170                                Résidus  Milliers de tonnes   
11          5142                             Nourriture  Milliers de tonnes   

    Valeur  
0    35798  
1     2349  
2    -1819  
3    20446  
4    19521  
5     7599  
6      708  
7      326  
8     1960  
9     1149  
10       0  
11    7779

In [10]:
data_ble_france = data_ble_france2018

In [11]:
#Le premier terme
terme1=data_ble_france[data_ble_france['Code Élément']==5301]['Valeur'].iloc[0] #valeur de disponibilité intérieur

In [12]:
#Le second terme
terme2=data_ble_france[data_ble_france['Code Élément']==5511]['Valeur'].iloc[0] + \
data_ble_france[data_ble_france['Code Élément']==5611]['Valeur'].iloc[0] + \
data_ble_france[data_ble_france['Code Élément']==5072]['Valeur'].abs() .iloc[0]- \
data_ble_france[data_ble_france['Code Élément']==5911]['Valeur'].iloc[0]

In [13]:
#Le troisième terme
terme3=data_ble_france[data_ble_france['Code Élément']==5521]['Valeur'].iloc[0] +\
    data_ble_france[data_ble_france['Code Élément']==5527]['Valeur'].iloc[0] + \
    data_ble_france[data_ble_france['Code Élément']==5123]['Valeur'].iloc[0]+ \
    data_ble_france[data_ble_france['Code Élément']==5131]['Valeur'].iloc[0] + \
    data_ble_france[data_ble_france['Code Élément']==5154]['Valeur'].iloc[0]+ \
    data_ble_france[data_ble_france['Code Élément']==5142]['Valeur'].iloc[0]

In [14]:
print('Terme1 = ',terme1)
print('Terme2 = ',terme2)
print('Terme2 = ',terme3)

Terme1 =  19521
Terme2 =  19520
Terme2 =  19521


In [15]:
# Selectionner une valeur : 
# data_ble_france[data_ble_france['Code Élément']==5511]['Valeur'].iloc[0]
# data_ble_france[data_ble_france['Code Élément']==5511]['Valeur'] .to_numpy()[0]

## Exemple blé en france 2013

In [17]:
data_france2013 = df_vegetaux[df_vegetaux['Code zone']==68]
data_ble_france2013= data_france2013[data_france2013['Code Produit']==2511]
#data_ble_france2013

In [18]:
data_ble_france2013Bis = df_vegetaux[ (df_vegetaux['Code zone']==68) & (df_vegetaux['Code Produit']==2511)]
#data_ble_france2013Bis

In [19]:
#Le premier terme
terme1=data_ble_france[data_ble_france['Code Élément']==5301]['Valeur'].iloc[0] #valeur de disponibilité intérieur

In [20]:
#Le second terme
terme2=data_ble_france[data_ble_france['Code Élément']==5511]['Valeur'].iloc[0] + \
data_ble_france[data_ble_france['Code Élément']==5611]['Valeur'].iloc[0] + \
data_ble_france[data_ble_france['Code Élément']==5072]['Valeur'].abs() .iloc[0]- \
data_ble_france[data_ble_france['Code Élément']==5911]['Valeur'].iloc[0]

In [21]:
#Le troisième terme
terme3=data_ble_france[data_ble_france['Code Élément']==5521]['Valeur'].iloc[0] +\
    data_ble_france[data_ble_france['Code Élément']==5527]['Valeur'].iloc[0] + \
    data_ble_france[data_ble_france['Code Élément']==5123]['Valeur'].iloc[0]+ \
    data_ble_france[data_ble_france['Code Élément']==5131]['Valeur'].iloc[0] + \
    data_ble_france[data_ble_france['Code Élément']==5154]['Valeur'].iloc[0]+ \
    data_ble_france[data_ble_france['Code Élément']==5142]['Valeur'].iloc[0]

In [22]:
print('Terme1 = ',terme1)
print('Terme2 = ',terme2)
print('Terme2 = ',terme3)

Terme1 =  19521
Terme2 =  19520
Terme2 =  19521


# Question 3 : Calculer pour chaque pays et chaque produit la disponibilité alimentaire en kcal puis en kg de protéines.

## Etape 1  : garder uniquement les colonnes importantes

In [23]:
df_population.head(3)

Code Domaine                                            Domaine  Code zone  \
0         FBSH  Bilans Alimentaire (Ancienne méthodologie et p...          2   
1         FBSH  Bilans Alimentaire (Ancienne méthodologie et p...        202   
2         FBSH  Bilans Alimentaire (Ancienne méthodologie et p...          3   

             Zone  Code Élément            Élément  Code Produit     Produit  \
0     Afghanistan           511  Population totale          2501  Population   
1  Afrique du Sud           511  Population totale          2501  Population   
2         Albanie           511  Population totale          2501  Population   

   Code année  Année           Unité  Valeur Symbole Description du Symbole  
0        2013   2013  1000 personnes   30552     NaN      Donnée officielle  
1        2013   2013  1000 personnes   52776     NaN      Donnée officielle  
2        2013   2013  1000 personnes    3173     NaN      Donnée officielle

In [24]:
df_population1 = df_population[['Code zone','Valeur','Unité']]

In [25]:
df_animaux1 = df_animaux [['Code zone', 'Zone', 'Code Élément', 'Élément', 'Code Produit', 'Produit','Valeur', 'Unité']]
df_animaux1 = df_animaux1.assign(Type = 'animal')

In [26]:
df_vegetaux1 =   df_vegetaux [['Code zone', 'Zone', 'Code Élément', 'Élément', 'Code Produit', 'Produit','Valeur', 'Unité']]
df_vegetaux1 = df_vegetaux1.assign(Type = 'vegetal')

## Etape 2  : concaténation entre df_animaux1 et df_vegetaux1

In [27]:
df_Alimentation=pd.concat([df_animaux1,df_vegetaux1])

In [28]:
print('df_animaux1.shape =\t', df_animaux1.shape , '\ndf_vegetaux2.shape =\t',
      df_vegetaux1.shape,'\ndf_Alimentation.shape =\t',df_Alimentation.shape)

df_animaux1.shape =	 (37165, 9) 
df_vegetaux2.shape =	 (104871, 9) 
df_Alimentation.shape =	 (142036, 9)


In [29]:
#df_Alimentation.groupby(['Zone']).mean()

In [30]:
#df_Alimentation.groupby(['Code zone', 'Code Élément', 'Code Produit']).mean()

## En SQL :
	SELECT *
	FROM df_Alimentation
	WHERE Code Élément = 674 OR Code Élément = 664 
    
	=> RR = df_Alimentation[ (df_Alimentation['Code Élément']==664) | (df_Alimentation['Code Élément']==674)]
    
    Sachant que :
        "674","Disponibilité de protéines en quantité (g/personne/jour)
        "664","Disponibilité alimentaire (Kcal/personne/jour)"

## Jointure à gauche entre df_Alimentation, df_population1

In [31]:
df_alimentation_population = df_Alimentation.merge(df_population1, on='Code zone', how="left", suffixes=('_produit','_population'))

In [32]:
df_alimentation_population.head(2)

Code zone         Zone  Code Élément                  Élément  \
0          2  Afghanistan          5511               Production   
1          2  Afghanistan          5611  Importations - Quantité   

   Code Produit           Produit  Valeur_produit       Unité_produit    Type  \
0          2731  Viande de Bovins           134.0  Milliers de tonnes  animal   
1          2731  Viande de Bovins             6.0  Milliers de tonnes  animal   

   Valeur_population Unité_population  
0              30552   1000 personnes  
1              30552   1000 personnes

### Disponibilité alimentaire en kcal 

In [33]:
DisponibiliteAlimenatireKcal = df_alimentation_population[ df_alimentation_population['Code Élément']==664]
DisponibiliteAlimenatireKcal.head()

Code zone         Zone  Code Élément  \
5           2  Afghanistan           664   
12          2  Afghanistan           664   
20          2  Afghanistan           664   
28          2  Afghanistan           664   
35          2  Afghanistan           664   

                                           Élément  Code Produit  \
5   Disponibilité alimentaire (Kcal/personne/jour)          2731   
12  Disponibilité alimentaire (Kcal/personne/jour)          2732   
20  Disponibilité alimentaire (Kcal/personne/jour)          2734   
28  Disponibilité alimentaire (Kcal/personne/jour)          2735   
35  Disponibilité alimentaire (Kcal/personne/jour)          2736   

                   Produit  Valeur_produit       Unité_produit    Type  \
5         Viande de Bovins            27.0  Kcal/personne/jour  animal   
12  Viande d'Ovins/Caprins            31.0  Kcal/personne/jour  animal   
20     Viande de Volailles             9.0  Kcal/personne/jour  animal   
28           Viande, Autre             1.0  Kcal/personne/jour  animal   
35        Abats Comestible             5.0  Kcal/personne/jour  animal   

    Valeur_population Unité_population  
5               30552   1000 personnes  
12              30552   1000 personnes  
20              30552   1000 personnes  
28              30552   1000 personnes  
35              30552   1000 personnes

In [34]:
DisponibiliteAlimenatireKcal = DisponibiliteAlimenatireKcal.assign(Nb_Kcal=DisponibiliteAlimenatireKcal['Valeur_produit']*DisponibiliteAlimenatireKcal['Valeur_population']*1000)
DisponibiliteAlimenatireKcal=DisponibiliteAlimenatireKcal[['Code zone', 'Zone', 'Code Produit', 'Produit', 'Valeur_produit', 'Unité_produit', 'Valeur_population','Unité_population', 'Nb_Kcal','Type']]
DisponibiliteAlimenatireKcal.head()

Code zone         Zone  Code Produit                 Produit  \
5           2  Afghanistan          2731        Viande de Bovins   
12          2  Afghanistan          2732  Viande d'Ovins/Caprins   
20          2  Afghanistan          2734     Viande de Volailles   
28          2  Afghanistan          2735           Viande, Autre   
35          2  Afghanistan          2736        Abats Comestible   

    Valeur_produit       Unité_produit  Valeur_population Unité_population  \
5             27.0  Kcal/personne/jour              30552   1000 personnes   
12            31.0  Kcal/personne/jour              30552   1000 personnes   
20             9.0  Kcal/personne/jour              30552   1000 personnes   
28             1.0  Kcal/personne/jour              30552   1000 personnes   
35             5.0  Kcal/personne/jour              30552   1000 personnes   

        Nb_Kcal    Type  
5   824904000.0  animal  
12  947112000.0  animal  
20  274968000.0  animal  
28   30552000.0  animal  
35  152760000.0  animal

In [35]:
#DisponibiliteAlimenatireKcal[['Nb_Kcal','Code zone']].groupby(['Code zone']).sum()

### Disponibilité alimentaire en kg de protéines

In [36]:
DisponibiliteAlimenatireProteine = df_alimentation_population[ df_alimentation_population['Code Élément']==674]
DisponibiliteAlimenatireProteine.head(12)

Code zone            Zone  Code Élément  \
6           2     Afghanistan           674   
13          2     Afghanistan           674   
21          2     Afghanistan           674   
29          2     Afghanistan           674   
36          2     Afghanistan           674   
43          2     Afghanistan           674   
50          2     Afghanistan           674   
58          2     Afghanistan           674   
68          2     Afghanistan           674   
79          2     Afghanistan           674   
88          2     Afghanistan           674   
99        202  Afrique du Sud           674   

                                              Élément  Code Produit  \
6   Disponibilité de protéines en quantité (g/pers...          2731   
13  Disponibilité de protéines en quantité (g/pers...          2732   
21  Disponibilité de protéines en quantité (g/pers...          2734   
29  Disponibilité de protéines en quantité (g/pers...          2735   
36  Disponibilité de protéines en quantité (g/pers...          2736   
43  Disponibilité de protéines en quantité (g/pers...          2740   
50  Disponibilité de protéines en quantité (g/pers...          2743   
58  Disponibilité de protéines en quantité (g/pers...          2737   
68  Disponibilité de protéines en quantité (g/pers...          2744   
79  Disponibilité de protéines en quantité (g/pers...          2848   
88  Disponibilité de protéines en quantité (g/pers...          2761   
99  Disponibilité de protéines en quantité (g/pers...          2731   

                   Produit  Valeur_produit    Unité_produit    Type  \
6         Viande de Bovins            1.89  g/personne/jour  animal   
13  Viande d'Ovins/Caprins            1.91  g/personne/jour  animal   
21     Viande de Volailles            0.87  g/personne/jour  animal   
29           Viande, Autre            0.17  g/personne/jour  animal   
36        Abats Comestible            0.77  g/personne/jour  animal   
43            Beurre, Ghee            0.03  g/personne/jour  animal   
50                   Crème            0.04  g/personne/jour  animal   
58  Graisses Animales Crue            0.01  g/personne/jour  animal   
68                   Oeufs            0.38  g/personne/jour  animal   
79      Lait - Excl Beurre            6.12  g/personne/jour  animal   
88      Poissons Eau Douce            0.02  g/personne/jour  animal   
99        Viande de Bovins            7.23  g/personne/jour  animal   

    Valeur_population Unité_population  
6               30552   1000 personnes  
13              30552   1000 personnes  
21              30552   1000 personnes  
29              30552   1000 personnes  
36              30552   1000 personnes  
43              30552   1000 personnes  
50              30552   1000 personnes  
58              30552   1000 personnes  
68              30552   1000 personnes  
79              30552   1000 personnes  
88              30552   1000 personnes  
99              52776   1000 personnes

In [37]:
DisponibiliteAlimenatireProteine = DisponibiliteAlimenatireProteine.assign(Nb_Kg_proteine=DisponibiliteAlimenatireProteine['Valeur_produit']*DisponibiliteAlimenatireProteine['Valeur_population'])
DisponibiliteAlimenatireProteine = DisponibiliteAlimenatireProteine[['Code zone', 'Zone', 'Code Produit', 'Produit', 'Valeur_produit', 'Unité_produit', 'Valeur_population','Unité_population', 'Nb_Kg_proteine','Type']]
DisponibiliteAlimenatireProteine.head()

Code zone         Zone  Code Produit                 Produit  \
6           2  Afghanistan          2731        Viande de Bovins   
13          2  Afghanistan          2732  Viande d'Ovins/Caprins   
21          2  Afghanistan          2734     Viande de Volailles   
29          2  Afghanistan          2735           Viande, Autre   
36          2  Afghanistan          2736        Abats Comestible   

    Valeur_produit    Unité_produit  Valeur_population Unité_population  \
6             1.89  g/personne/jour              30552   1000 personnes   
13            1.91  g/personne/jour              30552   1000 personnes   
21            0.87  g/personne/jour              30552   1000 personnes   
29            0.17  g/personne/jour              30552   1000 personnes   
36            0.77  g/personne/jour              30552   1000 personnes   

    Nb_Kg_proteine    Type  
6         57743.28  animal  
13        58354.32  animal  
21        26580.24  animal  
29         5193.84  animal  
36        23525.04  animal

In [38]:
#DisponibiliteAlimenatireProteine[['Nb_Kg_proteine','Code zone']].groupby(['Code zone']).sum()

# Réponse 4 : Ratio "énergie/poids" (en kcal/kg) pour chaque produit et chaque pays

### premier essai

In [39]:
# c=[]
# for pays in df_Alimentation['Code zone']:
#     for produit in df_Alimentation['Code Produit']:
#         a = df_Alimentation[df_Alimentation['Code Élément']==664]['Valeur']  /365
#         b = df_Alimentation[df_Alimentation['Code Élément']==645]['Valeur']
#         c.append(a/b)
        

### Seconde essai

In [40]:
# from time import time
# start_time = time()

# dico= {}

# zone_produit = df_Alimentation.groupby(['Code zone','Code Produit'])
# liste_groupes = zone_produit.groups


# for groupe in liste_groupes:
#     a = zone_produit.get_group(groupe)
#     kcal = a['Code Élément']==645
#     kg = a['Code Élément']==664

    
#     if kcal.any()   &  kg.any() :
#         if (a[kcal]['Valeur'].iloc[0]!= 0) :
#             dico[groupe]  = (a[kg]['Valeur'].iloc[0]  / 365 ) /  a[kcal]['Valeur'].iloc[0]
            

# end_time = time()
# time_taken = end_time - start_time # time_taken is in seconds

# print ("time_taken = ",time_taken)
# print(dico)

###  Solution : changer les " éléments " en colonnes puis garder uniquement 645 et 664 
       

### A. Supprimer les Nan avant de faire mon opération

In [41]:
df_rapport_calorique = pd.pivot_table (df_Alimentation, values='Valeur', index = ['Code zone','Code Produit'], 
                                columns = 'Code Élément')[[645,664]]
#df_rapport_calorique.dropna(inplace=True)
df_rapport_calorique = df_rapport_calorique[~df_rapport_calorique.isin([np.nan, np.inf, -np.inf, 0]).any(1)]
df_rapport_calorique['Ratio Energie/poids'] = df_rapport_calorique[664]*365 / df_rapport_calorique[645]
df_rapport_calorique['Unité du ratio'] ='kcal/kg'
df_rapport_calorique.head()

Code Élément               645     664  Ratio Energie/poids Unité du ratio
Code zone Code Produit                                                    
1         2511          130.60  1024.0          2861.868300        kcal/kg
          2515            0.12     1.0          3041.666667        kcal/kg
          2516            0.37     2.0          1972.972973        kcal/kg
          2520            1.96    14.0          2607.142857        kcal/kg
          2531           45.10    82.0           663.636364        kcal/kg

### B. Sans supprimer les Nan avant de faire mon opération

In [42]:
df_rapport_calorique2 = pd.pivot_table (df_Alimentation, values='Valeur', index = ['Code zone','Code Produit'], 
                                columns = 'Code Élément')[[645,664]]
df_rapport_calorique2['Ratio Energie/poids'] = df_rapport_calorique2[664]*365 / df_rapport_calorique2[645]
df_rapport_calorique2['Unité du ratio'] ='kcal/kg'
df_rapport_calorique2.head(5)

Code Élément               645     664  Ratio Energie/poids Unité du ratio
Code zone Code Produit                                                    
1         2511          130.60  1024.0          2861.868300        kcal/kg
          2513            0.00     0.0                  NaN        kcal/kg
          2514            0.03     0.0             0.000000        kcal/kg
          2515            0.12     1.0          3041.666667        kcal/kg
          2516            0.37     2.0          1972.972973        kcal/kg

In [43]:
print (df_rapport_calorique.shape)
print (df_rapport_calorique2.shape)

(9955, 4)
(15702, 4)


# Ratio "poids de protéines/poids total" 

In [44]:
df_rapport_proteine = pd.pivot_table (df_Alimentation, values='Valeur', index = ['Code zone','Code Produit'], 
                                columns = 'Code Élément')[[674,645]]
#df_rapport_proteine.dropna(inplace=True)
df_rapport_proteine = df_rapport_proteine[~df_rapport_proteine.isin([np.nan, np.inf, -np.inf, 0]).any(1)]
df_rapport_proteine['Ratio proteine/poids pour 100g'] = df_rapport_proteine[674]*365*10 / df_rapport_proteine[645]
df_rapport_proteine['unité']="g"

df_rapport_proteine


Code Élément              674     645  Ratio proteine/poids pour 100g unité
Code zone Code Produit                                                     
1         2511          30.52  130.60                      852.970904     g
          2514           0.01    0.03                     1216.666667     g
          2515           0.02    0.12                      608.333333     g
          2516           0.09    0.37                      887.837838     g
          2520           0.39    1.96                      726.275510     g
...                       ...     ...                             ...   ...
351       2767           0.56    8.85                      230.960452     g
          2769           0.07    0.67                      381.343284     g
          2775           0.69    9.24                      272.564935     g
          2805          14.47   77.45                      681.930278     g
          2848           3.12   33.18                      343.218807     g

[8841 rows x 4 columns]

In [45]:
a = df_rapport_proteine.index
a = a.droplevel(1)
a.drop_duplicates(keep="first")
a

Int64Index([  1,   1,   1,   1,   1,   1,   1,   1,   1,   1,
            ...
            351, 351, 351, 351, 351, 351, 351, 351, 351, 351],
           dtype='int64', name='Code zone', length=8841)

In [46]:
df_rapport_proteine.loc[(1,2516),:]

Code Élément
674                                     0.09
645                                     0.37
Ratio proteine/poids pour 100g    887.837838
unité                                      g
Name: (1, 2516), dtype: object

# Réponse 5 : Les 5 elements les plus caloriques

In [47]:
Elts_caloriques = df_rapport_calorique2.sort_values(by='Ratio Energie/poids', ascending=False)

In [48]:
Elts_caloriques

Code Élément            645  664  Ratio Energie/poids Unité du ratio
Code zone Code Produit                                              
67        2558          0.0  3.0                  inf        kcal/kg
169       2558          0.0  2.0                  inf        kcal/kg
19        2558          0.0  1.0                  inf        kcal/kg
84        2558          0.0  2.0                  inf        kcal/kg
107       2558          0.0  1.0                  inf        kcal/kg
...                     ...  ...                  ...            ...
351       2642          0.0  0.0                  NaN        kcal/kg
          2659          NaN  NaN                  NaN        kcal/kg
          2781          0.0  0.0                  NaN        kcal/kg
          2782          0.0  0.0                  NaN        kcal/kg
          2899          NaN  0.0                  NaN        kcal/kg

[15702 rows x 4 columns]

In [49]:
df_rapport_calorique2 = df_rapport_calorique2[~df_rapport_calorique2.isin([np.nan, np.inf, -np.inf, 0]).any(1)]

In [50]:
Elts_caloriques = df_rapport_calorique2.sort_values(by='Ratio Energie/poids', ascending=False)

In [51]:
Elts_caloriques.head(5)

Code Élément             645  664  Ratio Energie/poids Unité du ratio
Code zone Code Produit                                               
40        2586          0.05  3.0              21900.0        kcal/kg
117       2572          0.02  1.0              18250.0        kcal/kg
173       2582          0.02  1.0              18250.0        kcal/kg
81        2580          0.02  1.0              18250.0        kcal/kg
195       2580          0.02  1.0              18250.0        kcal/kg

# Moyenne

In [52]:
#df_rapport_calorique2.index.names
df_moyenne_rapport_calorique2 = df_rapport_calorique2.groupby(level="Code Produit").mean().sort_values(by='Ratio Energie/poids', ascending=False)
df_moyenne_rapport_calorique2.head(5)

Code Élément       645        664  Ratio Energie/poids
Code Produit                                          
2782          0.327500   8.500000         11809.003531
2579          0.249833   6.183333         10129.047094
2580          0.854109  20.689922          9314.953332
2581          0.308462   7.461538          9300.167741
2582          0.678381  16.409524          9260.445939

In [53]:
df_moyenne_rapport_calorique = df_rapport_calorique.groupby(level="Code Produit").mean().sort_values(by ='Ratio Energie/poids', ascending = False)
df_moyenne_rapport_calorique2.head(5)

Code Élément       645        664  Ratio Energie/poids
Code Produit                                          
2782          0.327500   8.500000         11809.003531
2579          0.249833   6.183333         10129.047094
2580          0.854109  20.689922          9314.953332
2581          0.308462   7.461538          9300.167741
2582          0.678381  16.409524          9260.445939

# Réponse 4 bis


In [54]:
df_ratio = pd.pivot_table (df_Alimentation, values='Valeur', index = ['Code zone','Zone','Code Produit', 'Produit','Type'], 
                                columns = ['Code Élément','Élément'])
#avoine = df_rapport_calorique.iloc[:,674]*365*1000 / df_rapport_calorique.iloc[:,645]

In [55]:
df_ratio.reset_index()

Code Élément Code zone     Zone Code Produit                   Produit  \
Élément                                                                  
0                    1  Arménie         2511                       Blé   
1                    1  Arménie         2513                      Orge   
2                    1  Arménie         2514                      Maïs   
3                    1  Arménie         2515                    Seigle   
4                    1  Arménie         2516                    Avoine   
...                ...      ...          ...                       ...   
15698              351    Chine         2781        Huiles de Poissons   
15699              351    Chine         2782  Huiles de Foie de Poisso   
15700              351    Chine         2805          Riz (Eq Blanchi)   
15701              351    Chine         2848        Lait - Excl Beurre   
15702              351    Chine         2899              Miscellanees   

Code Élément     Type                                                    645  \
Élément               Disponibilité alimentaire en quantité (kg/personne/an)   
0             vegetal                                             130.60       
1             vegetal                                               0.00       
2             vegetal                                               0.03       
3             vegetal                                               0.12       
4             vegetal                                               0.37       
...               ...                                                ...       
15698          animal                                               0.00       
15699          animal                                               0.00       
15700         vegetal                                              77.45       
15701          animal                                              33.18       
15702         vegetal                                                NaN       

Code Élément                                            664  \
Élément      Disponibilité alimentaire (Kcal/personne/jour)   
0                                                    1024.0   
1                                                       0.0   
2                                                       0.0   
3                                                       1.0   
4                                                       2.0   
...                                                     ...   
15698                                                   0.0   
15699                                                   0.0   
15700                                                 797.0   
15701                                                  59.0   
15702                                                   0.0   

Code Élément                                                      674  \
Élément      Disponibilité de protéines en quantité (g/personne/jour)   
0                                                         30.52         
1                                                          0.00         
2                                                          0.01         
3                                                          0.02         
4                                                          0.09         
...                                                         ...         
15698                                                      0.00         
15699                                                      0.00         
15700                                                     14.47         
15701                                                      3.12         
15702                                                      0.01         

Code Élément                                                           684  \
Élément      Disponibilité de matière grasse en quantité (g/personne/jour)   
0                                                          3.60       

In [56]:
#df_ratio = df_ratio[[664,645,674]]

df_ratio[11111,'Ratio Energie/poids'] = df_ratio[664,"Disponibilité alimentaire (Kcal/personne/jour)"]*365 / \
                        df_ratio[645,"Disponibilité alimentaire en quantité (kg/personne/an)"]
df_ratio.loc[:,(11112,'Unité du ratio')] ='kcal/kg'
df_ratio.head(20)


Code Élément                                                                                                   645    \
Élément                                                       Disponibilité alimentaire en quantité (kg/personne/an)   
Code zone Zone    Code Produit Produit                Type                                                             
1         Arménie 2511         Blé                    vegetal                                             130.60       
                  2513         Orge                   vegetal                                               0.00       
                  2514         Maïs                   vegetal                                               0.03       
                  2515         Seigle                 vegetal                                               0.12       
                  2516         Avoine                 vegetal                                               0.37       
                  2517         Millet                 vegetal                                                NaN       
                  2518         Sorgho                 vegetal                                                NaN       
                  2520         Céréales, Autres       vegetal                                               1.96       
                  2531         Pommes de Terre        vegetal                                              45.10       
                  2532         Manioc                 vegetal                                                NaN       
                  2534         Racines nda            vegetal                                                NaN       
                  2537         Sucre, betterave       vegetal                                                NaN       
                  2542         Sucre Eq Brut          vegetal                                              33.03       
                  2543         Edulcorants Autres     vegetal                                               1.82       
                  2546         Haricots               vegetal                                                NaN       
                  2547         Pois                   vegetal                                                NaN       
                  2549         Légumineuses Autres    vegetal                                               0.00       
                  2551         Noix                   vegetal                                               2.15       
                  2555         Soja                   vegetal                                               0.00       
                  2556         Arachides Decortiquees vegetal                                               0.40       

Code Élément                                                                                           664    \
Élément                                                       Disponibilité alimentaire (Kcal/personne/jour)   
Code zone Zone    Code Produit Produit                Type                                                     
1         Arménie 2511         Blé                    vegetal                                         1024.0   
                  2513         Orge                   vegetal                                            0.0   
                  2514         Maïs                   vegetal                                            0.0   
                  2515         Seigle                 vegetal                                            1.0   
                  2516         Avoine                 vegetal                                            2.0   
                  2517         Millet                 vegetal                                            NaN   
                  2518         Sorgho                 vegetal                                            NaN   
                  2520         Céréales, Autres       vegetal                                           14.0   
       

In [57]:
df_ratio.loc[[1]]                                      
df_ratio.xs(1, level=0, axis=0, drop_level=True)

Code Élément                                                                                           645    \
Élément                                               Disponibilité alimentaire en quantité (kg/personne/an)   
Zone    Code Produit Produit                  Type                                                             
Arménie 2511         Blé                      vegetal                                             130.60       
        2513         Orge                     vegetal                                               0.00       
        2514         Maïs                     vegetal                                               0.03       
        2515         Seigle                   vegetal                                               0.12       
        2516         Avoine                   vegetal                                               0.37       
...                                                                                                  ...       
        2781         Huiles de Poissons       animal                                                0.00       
        2782         Huiles de Foie de Poisso animal                                                0.00       
        2805         Riz (Eq Blanchi)         vegetal                                               3.18       
        2848         Lait - Excl Beurre       animal                                              209.03       
        2899         Miscellanees             vegetal                                                NaN       

Code Élément                                                                                   664    \
Élément                                               Disponibilité alimentaire (Kcal/personne/jour)   
Zone    Code Produit Produit                  Type                                                     
Arménie 2511         Blé                      vegetal                                         1024.0   
        2513         Orge                     vegetal                                            0.0   
        2514         Maïs                     vegetal                                            0.0   
        2515         Seigle                   vegetal                                            1.0   
        2516         Avoine                   vegetal                                            2.0   
...                                                                                              ...   
        2781         Huiles de Poissons       animal                                             0.0   
        2782         Huiles de Foie de Poisso animal                                             0.0   
        2805         Riz (Eq Blanchi)         vegetal                                           34.0   
        2848         Lait - Excl Beurre       animal                                           351.0   
        2899         Miscellanees             vegetal                                            3.0   

Code Élément                                                                                             674    \
Élément                                               Disponibilité de protéines en quantité (g/personne/jour)   
Zone    Code Produit Produit                  Type                                                               
Arménie 2511         Blé                      vegetal                                              30.52         
        2513         Orge                     vegetal                                               0.00         
        2514         Maïs                     vegetal                                               0.01         
        2515         Seigle                   vegetal                                               0.02         
        2516         Avoine                   vegetal                                               0.09         
...                                                           

In [58]:
# Multi indexing
#############################################################
# Afficher les informations de la zone 1 (sans la supprimer)#
# df_ratio.loc[[1]]                                         #
# df_ratio.xs(1, level=0, axis=0, drop_level=False)         #

#############################################################
#############################################################
# Afficher les informations de la zone 1 (sans la garder)   #
# df_ratio.loc[1]                                           #
# df_ratio.xs(1, level=0, axis=0, drop_level=True)          #
#############################################################

# df_ratio.query("Produit == 'Oeufs'")
df_ratio[df_ratio.index.get_level_values('Produit') == 'Oeufs'].head(5)


Code Élément                                                                                       645    \
Élément                                           Disponibilité alimentaire en quantité (kg/personne/an)   
Code zone Zone        Code Produit Produit Type                                                            
1         Arménie     2744         Oeufs   animal                                              11.69       
2         Afghanistan 2744         Oeufs   animal                                               1.30       
3         Albanie     2744         Oeufs   animal                                              12.45       
4         Algérie     2744         Oeufs   animal                                               8.06       
7         Angola      2744         Oeufs   animal                                               1.11       

Code Élément                                                                               664    \
Élément                                           Disponibilité alimentaire (Kcal/personne/jour)   
Code zone Zone        Code Produit Produit Type                                                    
1         Arménie     2744         Oeufs   animal                                           45.0   
2         Afghanistan 2744         Oeufs   animal                                            5.0   
3         Albanie     2744         Oeufs   animal                                           46.0   
4         Algérie     2744         Oeufs   animal                                           27.0   
7         Angola      2744         Oeufs   animal                                            4.0   

Code Élément                                                                                         674    \
Élément                                           Disponibilité de protéines en quantité (g/personne/jour)   
Code zone Zone        Code Produit Produit Type                                                              
1         Arménie     2744         Oeufs   animal                                               3.43         
2         Afghanistan 2744         Oeufs   animal                                               0.38         
3         Albanie     2744         Oeufs   animal                                               3.65         
4         Algérie     2744         Oeufs   animal                                               2.30         
7         Angola      2744         Oeufs   animal                                               0.33         

Code Élément                                                                                              684    \
Élément                                           Disponibilité de matière grasse en quantité (g/personne/jour)   
Code zone Zone        Code Produit Produit Type                                                                   
1         Arménie     2744         Oeufs   animal                                               3.15              
2         Afghanistan 2744         Oeufs   animal                                               0.36              
3         Albanie     2744         Oeufs   animal                                               3.28              
4         Algérie     2744         Oeufs   animal                                               1.85              
7         Angola      2744         Oeufs   animal                                               0.29              

Code Élément                                                   5072   5123   \
Élément                                           Variation de stock Pertes   
Code zone Zone        Code Produit Produit Type                               
1         Arménie     2744         Oeufs   animal                NaN    NaN   
2         Afghanistan 2744         Oeufs   animal                NaN    5.0   
3         Albanie     2744         Oeufs   animal                NaN    7.0   
4         Algérie     2744         Oeufs   animal         

In [59]:
df_ratio[11113,'Ratio protéine/poids pour 100g'] = df_ratio[674,"Disponibilité de protéines en quantité (g/personne/jour)"]*365*10 / \
                        df_ratio[645,"Disponibilité alimentaire en quantité (kg/personne/an)"]
df_ratio[11114,'Unité Ratio protéine/poids'] = 'g'

df_ratio[df_ratio.index.get_level_values('Produit') == 'Avoine'].head(5)


Code Élément                                                                                               645    \
Élément                                                   Disponibilité alimentaire en quantité (kg/personne/an)   
Code zone Zone               Code Produit Produit Type                                                             
1         Arménie            2516         Avoine  vegetal                                               0.37       
3         Albanie            2516         Avoine  vegetal                                               0.13       
4         Algérie            2516         Avoine  vegetal                                               0.01       
7         Angola             2516         Avoine  vegetal                                               0.01       
8         Antigua-et-Barbuda 2516         Avoine  vegetal                                               2.67       

Code Élément                                                                                       664    \
Élément                                                   Disponibilité alimentaire (Kcal/personne/jour)   
Code zone Zone               Code Produit Produit Type                                                     
1         Arménie            2516         Avoine  vegetal                                            2.0   
3         Albanie            2516         Avoine  vegetal                                            1.0   
4         Algérie            2516         Avoine  vegetal                                            0.0   
7         Angola             2516         Avoine  vegetal                                            0.0   
8         Antigua-et-Barbuda 2516         Avoine  vegetal                                           15.0   

Code Élément                                                                                                 674    \
Élément                                                   Disponibilité de protéines en quantité (g/personne/jour)   
Code zone Zone               Code Produit Produit Type                                                               
1         Arménie            2516         Avoine  vegetal                                               0.09         
3         Albanie            2516         Avoine  vegetal                                               0.03         
4         Algérie            2516         Avoine  vegetal                                               0.00         
7         Angola             2516         Avoine  vegetal                                               0.00         
8         Antigua-et-Barbuda 2516         Avoine  vegetal                                               0.64         

Code Élément                                                                                                      684    \
Élément                                                   Disponibilité de matière grasse en quantité (g/personne/jour)   
Code zone Zone               Code Produit Produit Type                                                                    
1         Arménie            2516         Avoine  vegetal                                               0.03              
3         Albanie            2516         Avoine  vegetal                                               0.01              
4         Algérie            2516         Avoine  vegetal                                               0.00              
7         Angola             2516         Avoine  vegetal                                               0.00              
8         Antigua-et-Barbuda 2516         Avoine  vegetal                                               0.24              

Code Élément                                                           5072   \
Élément                                                   Variation de stock   
Code zone Zone               Code Produit Produit Type                         
1         Arménie            2516   

#     Réponse 5 bis

In [60]:
df_ratio_energie_poids = df_ratio.loc[:,[11111,11112]]
df_ratio_energie_poids = df_ratio_energie_poids[~df_ratio_energie_poids.isin([np.nan, np.inf, -np.inf, 0]).any(1)]
df_moyenne_ratio_energie_poids = df_ratio_energie_poids.groupby(level = 'Code Produit').mean()
df_moyenne_ratio_energie_poids = df_moyenne_ratio_energie_poids.sort_values(by = (11111,'Ratio Energie/poids'), ascending = False)
df_moyenne_ratio_energie_poids.head(5)


Code Élément               11111
Élément      Ratio Energie/poids
Code Produit                    
2782                11809.003531
2579                10129.047094
2580                 9314.953332
2581                 9300.167741
2582                 9260.445939

In [61]:
df_ratio_proteine_poids = df_ratio.loc[:,[11113,11114]]
df_ratio_proteine_poids = df_ratio_proteine_poids[~df_ratio_proteine_poids.isin([np.nan, np.inf, -np.inf, 0]).any(1)]
df_moyenne_ratio_proteine_poids = df_ratio_proteine_poids.groupby(level = 'Code Produit').mean()
df_moyenne_ratio_proteine_poids = df_moyenne_ratio_proteine_poids.sort_values(by = (11113,'Ratio protéine/poids pour 100g'), ascending = False)
df_moyenne_ratio_proteine_poids.head(5)


Code Élément                          11113
Élément      Ratio protéine/poids pour 100g
Code Produit                               
2558                            3650.000000
2555                            2729.543759
2556                            2507.404775
2547                            2240.919309
2549                            2231.396481

#  Réponse 6 : Disponibilité intérieure mondiale pour les produits végétaux uniquement

In [62]:
df_ratio_vegetaux = df_ratio[df_ratio.index.get_level_values('Type') == 'vegetal']

In [63]:
df_ratio_vegetaux[[5301,]]

Code Élément                                                                      5301
Élément                                                       Disponibilité intérieure
Code zone Zone    Code Produit Produit                Type                            
1         Arménie 2511         Blé                    vegetal                    554.0
                  2513         Orge                   vegetal                    198.0
                  2514         Maïs                   vegetal                    102.0
                  2515         Seigle                 vegetal                      1.0
                  2516         Avoine                 vegetal                      6.0
...                                                                                ...
351       Chine   2659         Alcool, non Comestible vegetal                    239.0
                  2680         Aliments pour enfants  vegetal                    133.0
                  2745         Miel                   vegetal                    350.0
                  2805         Riz (Eq Blanchi)       vegetal                 135024.0
                  2899         Miscellanees           vegetal                      NaN

[11971 rows x 1 columns]

#  Réponse 6 BIS: Disponibilité intérieure mondiale en kcal pour les produits végétaux uniquement

In [64]:
df_vegetal_interieur = df_ratio.query("Type == 'vegetal'")[[5301,11111]]
df_vegetal_interieur.shape

(11971, 2)

In [65]:
df_vegetal_interieur1 = df_vegetal_interieur[~df_vegetal_interieur.isin([np.nan, np.inf, -np.inf, 0]).any(1)]
df_vegetal_interieur1.shape

(6766, 2)

In [66]:
df_vegetal_interieur1_kcal = df_vegetal_interieur1[(5301 , "Disponibilité intérieure")] * df_vegetal_interieur1[(11111 , "Ratio Energie/poids")] * 1000
df_vegetal_interieur1_kcal = df_vegetal_interieur1_kcal.groupby(level = 'Code Produit').sum()
df_vegetal_interieur1_kcal

Code Produit
2511    2.385296e+12
2513    2.704117e+11
2514    3.351612e+12
2515    4.835955e+10
2516    5.134236e+10
            ...     
2657    1.492520e+10
2658    1.036122e+11
2680    3.308634e+09
2745    6.048727e+09
2805    2.243327e+12
Length: 72, dtype: float64